plugins/modules/dcnm_image_upload.py

```diff
@@ -0,0 +1,777 @@
+
+        """
+        Routine to validate playbook input based on the state. Since each state has a different
+        config structure, this routine handles the validation based on the given state
+
+        Parameters:
+            None
+
+        Returns:
+            None
+        """
+
+        if [] is self.files:
+            return
+
+        cfg = []
+        for item in self.files:
+
+            citem = copy.deepcopy(item)
+
+            cfg.append(citem)
+
+            if self.module.params["state"] == "query":
+                # config for query state is different. So validate query state differently
+                self.dcnm_image_upload_validate_query_state_input(cfg)
+            elif self.module.params["state"] == "deleted":
+                # config for deleted state is different. So validate deleted state differently
+                self.dcnm_image_upload_validate_deleted_state_input(cfg)
+            else:
+                self.dcnm_image_upload_validate_input(cfg)
+            cfg.remove(citem)
+
+    def dcnm_image_upload_get_payload(self, image_upload_info):
+
+        """
+        This routine builds the complete object payload based on the information in self.want
+
+        Parameters:
+            image_upload_info (dict): Object information
+
+        Returns:
+            image_upload_payload (dict): Object payload information populated with appropriate data from playbook config
+        """
+
+        if image_upload_info["source"] != "local":
+            image_upload_payload = {
+                # Fill in the parameters that are required for creating/replacing image_upload object
+                "server": image_upload_info["remote_server"],
+                "filePath": image_upload_info["path"],
+                "userName": image_upload_info["username"],
+                "password": image_upload_info["password"],
+                "acceptHostKey": "false",
+                "source": image_upload_info["source"],
+            }
+        else:
+            image_upload_payload = {
+                # Fill in the parameters that are required for creating/replacing image_upload object
+                "filePath": image_upload_info["path"],
+                "source": image_upload_info["source"],
+            }
+
+        return image_upload_payload
+
+    def dcnm_image_upload_handle_local_file_transfer(self, elem):
+
+        """
+        Routine to read a local file specified in the playbook and transfer the file to DCNM controller
+
+        Parameters:
+            elem (dict): A dict containing complete path to local file
+
+        Returns:
+            True if file successfully transfered
+            False otherwise
+        """
+
+        protocol, address = dcnm_get_protocol_and_address(self.module)
+
+        path = protocol + ":" + address + self.paths["DCNM_CREATE_IMAGE_LOCAL"]
+
+        # We should use the authentication token already obtained to send data over this connection.
```

Comment:
Is this something we need to fix in our dcnm_send function?

Reply:
As explained in the meeting, we are using "requests" module to send multi-part-frames for uploading images. For using requests we should build auth headers and also the complete URL appropriately. So we cannot use dcnm_send here.